File: mobile/lib/domain/templates/fr_v1.dart
/// Template de réserve FR, version fr_v1 - miroir Dart de
/// `backend/app/domain/templates/fr_v1.py`.
///
/// Règles strictes appliquées ici (GATE zéro invention, section 2.4) :
///   - Aucun adjectif de gravité ("inutilisable", "dangereux", "détruit")
///     n'est généré par ce module ; seules les valeurs de
///     `packagingCondition` / `productCondition` telles que confirmées par
///     l'utilisateur sont citées (après passage par
///     `lib/domain/liability_guard.dart`, voir reserve_composer.dart).
///   - Aucune attribution de faute ou de responsabilité.
///   - Aucune mention de montant, indemnisation ou obligation juridique.
///   - Un champ marqué UNKNOWN par l'utilisateur reste explicitement signalé
///     comme non déterminé ; il n'est jamais tu ni inventé.
///
/// Pour une même entrée (ConfirmedFactData) et une même version de template,
/// la sortie est strictement identique (déterministe, pas d'appel LLM, pas
/// d'horodatage dans le texte) - IDENTIQUE bit-à-bit à la sortie du même
/// template côté backend (`backend/app/domain/templates/fr_v1.py`), ce qui
/// permet de vérifier une réserve générée hors ligne contre un recalcul
/// serveur ultérieur si besoin (section 6.1).
library;

import '../fact_set/confirmed_fact_data.dart';
import '../value_objects/issue_type.dart';

const String templateVersionFrV1 = 'fr_v1';

const Map<IssueType, String> _issueLabels = <IssueType, String>{
  IssueType.packagingDamage: "Dommage sur l'emballage",
  IssueType.productDamage: 'Dommage sur le produit',
  IssueType.missingQty: 'Quantité manquante',
  IssueType.wrongQty: 'Quantité incorrecte',
  IssueType.wrongProduct: 'Produit non conforme à la commande',
  IssueType.visibleNonconformity: 'Non-conformité visible',
  IssueType.other: 'Anomalie constatée',
};

const Map<String, String> _unknownFieldLabels = <String, String>{
  'productLabel': 'la désignation du produit',
  'productReference': 'la référence produit',
  'expectedQuantity': 'la quantité attendue',
  'receivedQuantity': 'la quantité reçue',
  'affectedQuantity': 'la quantité affectée',
  'packagingCondition': "l'état de l'emballage",
  'productCondition': "l'état du produit",
  'locationOnItem': "la localisation sur l'article",
};

String _productReferenceClause(ConfirmedFactData fact) {
  final List<String> parts = <String>[];
  if (fact.productLabel != null && fact.productLabel!.isNotEmpty) {
    parts.add(fact.productLabel!);
  }
  if (fact.productReference != null && fact.productReference!.isNotEmpty) {
    parts.add('(référence ${fact.productReference})');
  }
  if (parts.isEmpty) {
    return 'produit non identifié à ce stade';
  }
  return parts.join(' ');
}

String _fmtNum(double value) {
  if (value == value.roundToDouble()) {
    return value.toInt().toString();
  }
  // Équivalent de Python `f"{value:g}"` : nombre de décimales minimal.
  String text = value.toString();
  if (text.contains('.')) {
    text = text.replaceFirst(RegExp(r'0+$'), '');
    text = text.replaceFirst(RegExp(r'\.$'), '');
  }
  return text;
}

String _quantityClause(ConfirmedFactData fact) {
  final double? expected = fact.expectedQuantity;
  final double? received = fact.receivedQuantity;
  final double? missing = fact.missingQuantity; // calculé par code uniquement
  if (expected == null && received == null) {
    return 'quantités non précisées à ce stade';
  }
  if (expected != null && received != null) {
    String clause =
        'quantité attendue ${_fmtNum(expected)}, quantité reçue ${_fmtNum(received)}';
    if (missing != null) {
      clause += ', écart constaté de ${_fmtNum(missing)}';
    }
    return clause;
  }
  if (expected != null) {
    return 'quantité attendue ${_fmtNum(expected)}, quantité reçue non précisée';
  }
  return 'quantité reçue ${_fmtNum(received!)}, quantité attendue non précisée';
}

String? _conditionClause(ConfirmedFactData fact) {
  final List<String> clauses = <String>[];
  if (fact.packagingCondition != null && fact.packagingCondition!.isNotEmpty) {
    clauses.add('emballage constaté : ${fact.packagingCondition}');
  }
  if (fact.productCondition != null && fact.productCondition!.isNotEmpty) {
    clauses.add('produit constaté : ${fact.productCondition}');
  }
  if (clauses.isEmpty) {
    return null;
  }
  return clauses.join('; ');
}

String? _unknownClause(ConfirmedFactData fact) {
  if (fact.unknownFields.isEmpty) {
    return null;
  }
  final List<String> labels = fact.unknownFields
      .map((String f) => _unknownFieldLabels[f] ?? f)
      .toList()
    ..sort();
  return 'Non déterminé à ce stade : ${labels.join(', ')}.';
}

String _capitalize(String value) {
  if (value.isEmpty) {
    return value;
  }
  return value[0].toUpperCase() + value.substring(1);
}

/// Compose le paragraphe déterministe pour UNE anomalie (un
/// ConfirmedFactData).
///
/// Précondition : `fact` a nécessairement `userConfirmed == true` (invariant
/// de type, voir `ConfirmedFactData`) ; ce module ne revalide pas cet
/// invariant. Précondition additionnelle depuis R0.1 : `fact` a déjà été
/// filtré par `lib/domain/liability_guard.dart::screenConfirmedFact` (fait
/// par `reserve_composer.composeReserve`, jamais par ce module directement).
String composeIssueParagraph(ConfirmedFactData fact) {
  final String label = _issueLabels[fact.issueType]!;
  final List<String> sentences = <String>[
    '$label - ${_productReferenceClause(fact)}.',
  ];

  if (fact.issueType == IssueType.missingQty || fact.issueType == IssueType.wrongQty) {
    sentences.add('${_capitalize(_quantityClause(fact))}.');
  }

  final String? condition = _conditionClause(fact);
  if (condition != null) {
    sentences.add('${_capitalize(condition)}.');
  }

  if (fact.locationOnItem != null && fact.locationOnItem!.isNotEmpty) {
    sentences.add('Localisation : ${fact.locationOnItem}.');
  }

  if (fact.affectedQuantity != null) {
    sentences.add('Quantité affectée : ${_fmtNum(fact.affectedQuantity!)}.');
  }

  final String? unknown = _unknownClause(fact);
  if (unknown != null) {
    sentences.add(unknown);
  }

  if (fact.userUncertainty) {
    sentences.add("L'utilisateur signale une incertitude résiduelle sur ces constats.");
  }

  return sentences.join(' ');
}

const String prudenceMentionFrV1 =
    'ReserveFlash vous aide à structurer vos constats et vos preuves. '
    "Il ne remplace pas un conseil juridique. Vérifiez les conditions applicables "
    'à votre transport et à vos contrats.';

File: backend/app/domain/templates/fr_v1.py
"""Template de réserve FR, version fr_v1 (section 2.4 - "Templates versionnés
par langue et type d'incident").

Règles strictes appliquées ici (GATE zéro invention, section 2.4) :
  - Aucun adjectif de gravité ("inutilisable", "dangereux", "détruit") n'est
    généré par ce module ; seules les valeurs de `packaging_condition` /
    `product_condition` telles que confirmées par l'utilisateur sont citées.
  - Aucune attribution de faute ou de responsabilité.
  - Aucune mention de montant, indemnisation ou obligation juridique.
  - Un champ marqué UNKNOWN par l'utilisateur reste explicitement signalé
    comme non déterminé ; il n'est jamais tu ni inventé.

Pour une même entrée (ConfirmedFactData) et une même version de template, la
sortie est strictement identique (déterministe, pas d'appel LLM, pas
d'horodatage dans le texte).
"""

from __future__ import annotations

from app.domain.fact_set import ConfirmedFactData
from app.domain.value_objects import IssueType

TEMPLATE_VERSION = "fr_v1"

_ISSUE_LABELS: dict[IssueType, str] = {
    IssueType.PACKAGING_DAMAGE: "Dommage sur l'emballage",
    IssueType.PRODUCT_DAMAGE: "Dommage sur le produit",
    IssueType.MISSING_QTY: "Quantité manquante",
    IssueType.WRONG_QTY: "Quantité incorrecte",
    IssueType.WRONG_PRODUCT: "Produit non conforme à la commande",
    IssueType.VISIBLE_NONCONFORMITY: "Non-conformité visible",
    IssueType.OTHER: "Anomalie constatée",
}

_UNKNOWN_FIELD_LABELS: dict[str, str] = {
    "product_label": "la désignation du produit",
    "product_reference": "la référence produit",
    "expected_quantity": "la quantité attendue",
    "received_quantity": "la quantité reçue",
    "affected_quantity": "la quantité affectée",
    "packaging_condition": "l'état de l'emballage",
    "product_condition": "l'état du produit",
    "location_on_item": "la localisation sur l'article",
}


def _product_reference_clause(fact: ConfirmedFactData) -> str:
    parts: list[str] = []
    if fact.product_label:
        parts.append(fact.product_label)
    if fact.product_reference:
        parts.append(f"(référence {fact.product_reference})")
    if not parts:
        return "produit non identifié à ce stade"
    return " ".join(parts)


def _quantity_clause(fact: ConfirmedFactData) -> str:
    expected = fact.expected_quantity
    received = fact.received_quantity
    missing = fact.missing_quantity  # calculé par code, jamais par le LLM
    if expected is None and received is None:
        return "quantités non précisées à ce stade"
    if expected is not None and received is not None:
        clause = f"quantité attendue {_fmt_num(expected)}, quantité reçue {_fmt_num(received)}"
        if missing is not None:
            clause += f", écart constaté de {_fmt_num(missing)}"
        return clause
    if expected is not None:
        return f"quantité attendue {_fmt_num(expected)}, quantité reçue non précisée"
    return f"quantité reçue {_fmt_num(received)}, quantité attendue non précisée"


def _fmt_num(value: float) -> str:
    if float(value).is_integer():
        return str(int(value))
    return f"{value:g}"


def _condition_clause(fact: ConfirmedFactData) -> str | None:
    clauses: list[str] = []
    if fact.packaging_condition:
        clauses.append(f"emballage constaté : {fact.packaging_condition}")
    if fact.product_condition:
        clauses.append(f"produit constaté : {fact.product_condition}")
    if not clauses:
        return None
    return "; ".join(clauses)


def _unknown_clause(fact: ConfirmedFactData) -> str | None:
    if not fact.unknown_fields:
        return None
    labels = [_UNKNOWN_FIELD_LABELS.get(f, f) for f in fact.unknown_fields]
    return "Non déterminé à ce stade : " + ", ".join(sorted(labels)) + "."


def compose_issue_paragraph(fact: ConfirmedFactData) -> str:
    """Compose le paragraphe déterministe pour UNE anomalie (un ConfirmedFactData).

    Précondition : fact.user_confirmed est déjà garanti True par le modèle
    pydantic (voir ConfirmedFactData._enforce_user_confirmed) ; ce module ne
    revalide pas cet invariant, il fait confiance au type.
    """
    label = _ISSUE_LABELS[fact.issue_type]
    sentences: list[str] = [f"{label} - {_product_reference_clause(fact)}."]

    if fact.issue_type in (IssueType.MISSING_QTY, IssueType.WRONG_QTY):
        sentences.append(_quantity_clause(fact).capitalize() + ".")

    condition = _condition_clause(fact)
    if condition:
        sentences.append(condition.capitalize() + ".")

    if fact.location_on_item:
        sentences.append(f"Localisation : {fact.location_on_item}.")

    if fact.affected_quantity is not None:
        sentences.append(f"Quantité affectée : {_fmt_num(fact.affected_quantity)}.")

    unknown = _unknown_clause(fact)
    if unknown:
        sentences.append(unknown)

    if fact.user_uncertainty:
        sentences.append("L'utilisateur signale une incertitude résiduelle sur ces constats.")

    return " ".join(sentences)


PRUDENCE_MENTION = (
    "ReserveFlash vous aide à structurer vos constats et vos preuves. "
    "Il ne remplace pas un conseil juridique. Vérifiez les conditions applicables "
    "à votre transport et à vos contrats."
)
